From 7d3906048a36c545cbd7e6077ed1c534e6622aea Mon Sep 17 00:00:00 2001
From: Matheus Streb <matheus.streb@nubo.solutions>
Date: Tue, 7 Mar 2017 12:55:54 -0300
Subject: [PATCH 1/1] Added a new setting for color blind users (turn on and
 off) Added a customizable rules on less for color-blind status Added a
 example of usage at accountPreferences view

---
 packages/rocketchat-i18n/i18n/pt.i18n.json                   |  2 ++
 .../app/client/stylesheets/utils/_variables.import.less      |  7 +++++++
 .../rocketchat-ui-account/client/accountPreferences.coffee   |  1 +
 .../rocketchat-ui-account/client/accountPreferences.html     | 12 ++++++++++++
 packages/rocketchat-ui-sidenav/client/accountBox.coffee      |  6 +++++-
 server/methods/saveUserPreferences.js                        |  5 +++++
 6 files changed, 32 insertions(+), 1 deletion(-)

diff --git a/packages/rocketchat-i18n/i18n/pt.i18n.json b/packages/rocketchat-i18n/i18n/pt.i18n.json
index 18fd22c..bc241db 100644
--- a/packages/rocketchat-i18n/i18n/pt.i18n.json
+++ b/packages/rocketchat-i18n/i18n/pt.i18n.json
@@ -14,6 +14,7 @@
   "Accept_with_no_online_agents": "Aceitar sem agentes online",
   "Access_not_authorized": "Acesso não autorizado",
   "Access_Token_URL": "URL do Token de Acesso",
+  "Accessibility": "Acessibilidade",
   "Accessing_permissions": "Acessando permissões",
   "Account_SID": "Account SID",
   "Accounts": "Contas",
@@ -270,6 +271,7 @@
   "Closed_by_visitor": "Encerrado pelo visitante",
   "Closing_chat": "Encerrando chat",
   "Collapse_Embedded_Media_By_Default": "Esconder mídia por padrão",
+  "Color_Blind": "Daltônico",
   "Color": "Cor",
   "Commands": "Comandos",
   "Comment_to_leave_on_closing_session": "Comentário ao fechar sessão",
diff --git a/packages/rocketchat-livechat/app/client/stylesheets/utils/_variables.import.less b/packages/rocketchat-livechat/app/client/stylesheets/utils/_variables.import.less
index 27e8fd7..15e1417 100644
--- a/packages/rocketchat-livechat/app/client/stylesheets/utils/_variables.import.less
+++ b/packages/rocketchat-livechat/app/client/stylesheets/utils/_variables.import.less
@@ -30,3 +30,10 @@
 @status-away: #fcb316;
 
 @window-border-color: #e7e7e7;
+
+// Color Blind
+// Todo: choose the right color for color-blind users
+@status-online-color-blind: #4dff4d;
+@status-offline-color-blind: rgba(150, 150, 150, 0.5);
+@status-busy-color-blind: #d30230;
+@status-away-color-blind: #fcb316;
diff --git a/packages/rocketchat-ui-account/client/accountPreferences.coffee b/packages/rocketchat-ui-account/client/accountPreferences.coffee
index 5ed2003..e58c62a 100644
--- a/packages/rocketchat-ui-account/client/accountPreferences.coffee
+++ b/packages/rocketchat-ui-account/client/accountPreferences.coffee
@@ -97,6 +97,7 @@ Template.accountPreferences.onCreated ->
 		data.highlights = _.compact(_.map($('[name=highlights]').val().split(','), (e) -> return _.trim(e)))
 		data.desktopNotificationDuration = $('input[name=desktopNotificationDuration]').val()
 		data.unreadAlert = $('#unreadAlert').find('input:checked').val()
+		data.colorBlind = $('#colorBlind').find('input:checked').val()
 
 		Meteor.call 'saveUserPreferences', data, (error, results) ->
 			if results
diff --git a/packages/rocketchat-ui-account/client/accountPreferences.html b/packages/rocketchat-ui-account/client/accountPreferences.html
index c0c5f89..14188f9 100644
--- a/packages/rocketchat-ui-account/client/accountPreferences.html
+++ b/packages/rocketchat-ui-account/client/accountPreferences.html
@@ -214,6 +214,18 @@
 							</div>
 						</div>
 					</div>
+					<div class="section">
+						<h1>{{_ "Accessibility"}}</h1>
+						<div class="section-content border-component-color">
+							<div class="input-line double-col" id="colorBlind">
+								<label for="language">{{_ "Color_Blind"}}</label>
+								<div>
+									<label><input type="radio" name="colorBlind" value="1" checked="{{checked 'colorBlind' true }}" /> {{_ "True"}}</label>
+									<label><input type="radio" name="colorBlind" value="0" checked="{{checked 'colorBlind' false true}}" /> {{_ "False"}}</label>
+								</div>
+							</div>
+						</div>
+					</div>
 				</fieldset>
 				<div class="submit">
 					<button class="button primary save"><i class="icon-send"></i><span>{{_ "Save_changes"}}</span></button>
diff --git a/packages/rocketchat-ui-sidenav/client/accountBox.coffee b/packages/rocketchat-ui-sidenav/client/accountBox.coffee
index b0f2bf2..88a2934 100644
--- a/packages/rocketchat-ui-sidenav/client/accountBox.coffee
+++ b/packages/rocketchat-ui-sidenav/client/accountBox.coffee
@@ -2,6 +2,8 @@ Template.accountBox.helpers
 	myUserInfo: ->
 		visualStatus = "online"
 		username = Meteor.user()?.username
+		colorBlind = Meteor.user()?.settings?.preferences?.colorBlind
+		status = Session.get('user_' + username + '_status')
 		switch Session.get('user_' + username + '_status')
 			when "away"
 				visualStatus = t("away")
@@ -11,12 +13,14 @@ Template.accountBox.helpers
 				visualStatus = t("invisible")
 		return {
 			name: Session.get('user_' + username + '_name')
-			status: Session.get('user_' + username + '_status')
+			status: if colorBlind then status + '-color-blind' else status
 			visualStatus: visualStatus
 			_id: Meteor.userId()
 			username: username
 		}
 
+		
+
 	showAdminOption: ->
 		return RocketChat.authz.hasAtLeastOnePermission( ['view-statistics', 'view-room-administration', 'view-user-administration', 'view-privileged-setting' ]) or RocketChat.AdminBox.getOptions().length > 0
 
diff --git a/server/methods/saveUserPreferences.js b/server/methods/saveUserPreferences.js
index 2d951fd..010a5cc 100644
--- a/server/methods/saveUserPreferences.js
+++ b/server/methods/saveUserPreferences.js
@@ -55,6 +55,11 @@ Meteor.methods({
 				preferences.unreadAlert = settings.unreadAlert === '1' ? true : false;
 			}
 
+			if (settings.colorBlind) {
+				preferences.colorBlind = settings.colorBlind === '1' ? true : false;
+			}
+
+
 			preferences.desktopNotificationDuration = settings.desktopNotificationDuration - 0;
 			preferences.viewMode = settings.viewMode || 0;
 			preferences.hideUsernames = settings.hideUsernames === '1';
-- 
1.9.1

